packages/patterns/src/keys/compareKeys.js

```diff
@@ -6,13 +6,83 @@ import {
   compareRank,
   recordNames,
   recordValues,
+  trivialComparator,
 } from '@endo/marshal';
-import { assertKey } from './checkKey.js';
-import { bagCompare } from './merge-bag-operators.js';
-import { setCompare } from './merge-set-operators.js';
+import {
+  assertKey,
+  getCopyBagEntries,
+  getCopyMapEntries,
+  getCopySetKeys,
+} from './checkKey.js';
+import { makeCompareCollection } from './keycollection-operators.js';
+
+/** @template {Key} [K=Key] @typedef {import('../types').CopySet<K>} CopySet */
 
 const { quote: q, Fail } = assert;
 
+/**
+ * CopySet X is smaller than CopySet Y iff all of these conditions hold:
+ * 1. For every x in X, x is also in Y.
+ * 2. There is a y in Y that is not in X.
+ *
+ * X is equivalent to Y iff the condition 1 holds but condition 2 does not.
+ */
+export const setCompare = makeCompareCollection(
+  /** @type {<K extends unknown>(s: CopySet<K>) => Array<[K, 1]>} */ (
+    s => harden(getCopySetKeys(s).map(key => [key, 1]))
+  ),
+  0,
+  trivialComparator,
+);
+harden(setCompare);
+
+/**
+ * CopyBag X is smaller than CopyBag Y iff all of these conditions hold
+ * (where `count(A, a)` is shorthand for the count associated with `a` in `A`):
+ * 1. For every x in X, x is also in Y and count(X, x) <= count(Y, x).
+ * 2. There is a y in Y such that y is not in X or count(X, y) < count(Y, y).
+ *
+ * X is equivalent to Y iff the condition 1 holds but condition 2 does not.
+ */
+export const bagCompare = makeCompareCollection(
+  getCopyBagEntries,
+  0n,
+  trivialComparator,
+);
+harden(bagCompare);
+
+// TODO The desired semantics for CopyMap comparison have not yet been decided.
+// See https://github.com/endojs/endo/pull/1737#pullrequestreview-1596595411
+// The below is a currently-unused extension of CopyBag semantics (i.e., absent
+// entries treated as present with a value that is smaller than everything).
+const ABSENT = Symbol('absent');
```

Comment:
Worth mentioning why `Symbol.for('absent')` would not work ;)

Reply:
Good idea; done.